framework/include/base/libMeshReducedNamespace.h

```diff
@@ -0,0 +1,418 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+// This is the minimal set of headers to make all the types below that cannot be
+// forward-declared as classes available.
+// If we want to reduce this list, we would have to forward-declare functions as well
+#include "libmesh/libmesh.h"
+#include "libmesh/id_types.h"
+#include "libmesh/int_range.h"
+#include "libmesh/vector_value.h"
+#include "libmesh/tensor_value.h"
```

Comment:
I don't understand some of these includes, e.g. these. A comment about why they are included would be helpful

Reply:
this one brings a bunch of typedef so it must be why
```
typedef TensorValue<Real>   RealTensorValue;
typedef TensorValue<Number> NumberTensorValue;
typedef RealTensorValue     RealTensor;
typedef NumberTensorValue   Tensor;
```
there is no forward-declaration of typedefs. If I want to put RealTensor in the local namespace (using libMesh::RealTensor), I need to either add the typedef again or include the header.
including the header is better to reduce code duplication

Reply:
I ll double check these headers and separate them in two lists:
- needed to get a function without re-declaring it
- needed to get a typedef without re-defining it

Reply:
Relying on a forward declaration instead of a header from a library can be dangerous, since it turns some internal library refactoring (e.g. adding a new optional template argument or changing a typedef) into API-breaking refactoring that would otherwise have been fine.  I think it falls somewhere between "heavily discouraged" and "non-standards-conforming" for `std::` classes for this reason, with the standard going so far as to create `<iosfwd>` for the one case where people were *really* ticked about the long compile times of the full header.

For libMesh usage in MOOSE, on the other hand, I'd say it's fine to do forward declarations.  If we ever break compatibility with them then we'll fix it at submodule upgrade time.

On the gripping hand*, the only real benefit of forward declarations over header inclusion is that they reduce compile time, and I don't think the reduction is a large one for unity builds where every header is likely to end up included from another .C file anyway.  So either way is fine for us.

\* did I just use that idiom twice in one day? feels like I'm catching up for a long stretch of underuse...

Reply:
Yea Im glad I got to see two gripping hands today. We have made forward declaration headers in quite a few places throughout our software stack to try and help with changing class template signatures, etc. But anyway I think what's here is fine. Marking this as resolved

Reply:
imo we could chose to avoid tensor_value.h and vector_value.h as the typedefs are simple.
But for the ranges the typedefs are complicated and I think we even need to include other heads to be able to define them.